File: Data Structures & Algorithms/maximum-subarray/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    maxSubArray(nums) {
        // SLIDING WINDOW - prefix sum

        // init sum to MIN INT
        let res = Number.MIN_SAFE_INTEGER;
        let sum = 0;

        // run loop from i=0 to i<nums.length
        for(let i=0;i<nums.length;i++){
            // if nums[i] > sum
            if(sum < 0){
                // sum = nums[i]
                sum = 0;
            }
            // add nums[i] to sum
            sum+=nums[i];
            res = Math.max(res,sum)
        }

        // return sum
        return res;
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> res(nums.size());

        for(int i=0;i<nums.size();i++){
            res[i] = nums[i] * (i==0?1:res[i-1]);
        }

        for(int j=nums.size()-1;j>=0;j--){
            int postfix = nums[j] * (j==nums.size()-1?1:nums[j+1]);
            nums[j] = postfix;
            // res[j] = prefix * postfix;
            res[j] = (j==0?1:res[j-1]) * (j==nums.size()-1?1:nums[j+1]);
        }

        for(int k=0;k<nums.size();k++){
            cout<<res[k]<<" ";
        }

        return res;
    }
};
